_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,565 @@
+Another option is to separately configure TLS for connection encryption and integrity, but use a specific SASL protocol for authentication.
+
+There are several SASL based protocols you can use.
+One option is to use SASL PLAIN to authenticate with username and password.
+Another option is to use SASL SCRAM which is also username and password based, but rather than communicating, and storing clear text passwords, salted hashes are used.
+These two implementations both store usernames and some form of credentials in ZooKeeper.
+They require secure connection between Kafka clients and brokers, but also between Kafka brokers and ZooKeeper nodes.
+
+A more advanced, enterprise-ready solution is to use SASL GSSAPI mechanism which provides support for Kerberos. 
+That typically uses a central LDAP server for managing users, and permissions.
+The problem is that it is quite complex to setup, and people often find it hard to configure.
+
+Since Kafka version 2.0.0 there is an extensible OAuth 2.0 compatible token-based mechanism available, called SASL OAUTHBEARER.
+
+We have developed extensions that provide integration with OAuth 2.0 compliant authorization servers. 
+That means that in principle you can use any OAuth 2.0 compliant authorization server to expose centrally managed users for authentication with Kafka.
+
+
+## Why OAuth 2.0?
+
+Originally developed for scenarios where web users need to give limited access to their data inside one web site to another web site, OAuth 2.0 provides at least two main benefits.
+One is that user accounts and credentials for many services are managed centrally, organization-wide, by a dedicated identity service.
+Another important benefit is that the user only ever passes credentials like username and password to that dedicated server.
+The result of authenticating is a time-limited token which the user passes around to other services, and which serves to identify user, and to establish which permissions they have.
+ 
+
+## Hands-on
+
+Let's see how this works. As our OAuth 2.0 compliant authorization server we will use a basic setup of Keycloak with TLS support.
+We'll then configure a Kafka cluster with OAuth 2.0 support, and an example Kafka producer, and Kafka consumer client applications, that will use OAuth 2.0 to authenticate. 
+
+We'll assume that you have a Kubernetes cluster available in order to explain how this works for Strimzi, but the OAuth 2.0 support can be used with non-Strimzi Kafka installations too, by using [Strimzi Kafka OAuth libraries](https://github.com/strimzi/strimzi-kafka-oauth).
+ 
+If you are using `minikube` on your local machine, make sure to start `minikube tunnel` in another terminal in order for LoadBalancer services to get their externals IPs assigned:
+
+    sudo minikube tunnel 
+
+If you're wondering, an example of the tested `minikube` configuration is to run it as:
+
+    minikube start --vm-driver hyperkit --memory=4096 --cpus=4
```

Comment:
I doubt that we should have the `--vm-driver` in the starting command because for example for me it's using `kvm`